src/predictor/gpu_predictor.cu

```diff
@@ -237,6 +237,19 @@ class GPUPredictor : public xgboost::Predictor {
     }
   }
 
+  void PredictionDeviceOffsets(size_t total_size, size_t batch_offset, size_t batch_size,
+                               int n_classes, std::vector<size_t>* out_offsets) {
+    auto& offsets = *out_offsets;
+    size_t n_shards = devices_.Size();
+    offsets.resize(n_shards + 2);
+    size_t rows_per_shard = (batch_size + n_shards - 1) / n_shards;
+    for (int shard = 0; shard < devices_.Size(); ++shard) {
+      offsets[shard] = batch_offset + shard * rows_per_shard * n_classes;
+    }
+    offsets[n_shards] = batch_offset + batch_size * n_classes;
+    offsets[n_shards + 1] = total_size;
```

Comment:
Doesn't this mean that the last device gets a disproportionately large number of rows? Won't it lead to out-of-memory errors?

Reply:
I'm actually creating a vector of size `n_shards + 2`, so the last shard ends at the second to last element.

Reply:
I've looked into it in more detail. This actually satisfies the check that the last element of the offsets must be equal to the size of the HostDeviceVector. The check that `CHECK_EQ(offsets.size() + 1, devices.Size())` has been implied, but apparently never inserted. Your approach enables to avoid having GPU memory corresponding to the tail of HostDeviceVector.

This looks a bit hacky to me. However, if you decide to proceed with this option, please document this in `HostDeviceVector`, and add a test so that this option is not broken.

A more general alternative would be to have only a window of the whole HostDeviceVector visible on the GPU. However, this would make the implementation even more complicated.

Reply:
Yeah I was thinking we could add a `Windowed` distribution to `GPUDistribution` to support sliding over a vector from GPUs. Would that make sense? Probably in a followup PR.

Reply:
`Windowed` would go against 1:1 (mostly) correspondence between the contents on the host and on the GPU, which is something we'd like to keep.

Instead, something like a block-cyclic distribution may be better.

However, please add documentation to HostDeviceVector or GPUDistribution that you are using the case where `offsets.size() + 1 != devices.Size()`.

Reply:
Added a note to `GPUDistribution::Explicit()`.